identity_iota/src/io/writer.rs

```diff
@@ -0,0 +1,189 @@
+
+/// Fixed-address used for faster transaction confirmation times
+const PROMOTION: &str = "PROMOTEADDRESSPROMOTEADDRESSPROMOTEADDRESSPROMOTEADDRESSPROMOTEADDRESSPROMOTEADDR";
+
+#[derive(Debug)]
+pub struct Config {
+    confirm_delay: Duration,
+    promote_delay: Duration,
+    promote_limit: usize,
+}
+
+impl Config {
+    const DEFAULT_CONFIRM_DELAY: Duration = Duration::from_secs(5);
+    const DEFAULT_PROMOTE_DELAY: Duration = Duration::from_secs(5);
+    const DEFAULT_PROMOTE_LIMIT: usize = 20;
+
+    pub const fn new() -> Self {
+        Self {
+            confirm_delay: Self::DEFAULT_CONFIRM_DELAY,
+            promote_delay: Self::DEFAULT_PROMOTE_DELAY,
+            promote_limit: Self::DEFAULT_PROMOTE_LIMIT,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TangleWriter {
+    client: iota::Client,
+    config: Config,
+    network: Network,
+}
+
+impl TangleWriter {
+    pub fn new(nodelist: &NodeList) -> Result<Self> {
+        let mut builder: iota::ClientBuilder = iota::ClientBuilder::new();
+
+        for node in nodelist.nodes() {
+            builder = builder.node(node)?;
+        }
+
+        builder = builder.network(nodelist.network().into());
+
+        Ok(Self {
+            client: builder.build()?,
+            network: nodelist.network(),
+            config: Config::new(),
+        })
+    }
+
+    pub async fn write_json<T>(&self, did: &DID, data: &T) -> Result<Hash>
+    where
+        T: Serialize,
+    {
+        let txn: Hash = self.publish_json(did, data).await?;
+
+        println!("[+] Transaction > {}", encode_trits(txn.as_trits()));
+
+        let mut tries: usize = 0;
+
+        thread::sleep(self.config.confirm_delay);
+
+        while !self.is_confirmed(&txn).await? {
+            println!("[+] Promoting > {}", encode_trits(txn.as_trits()));
+
+            tries += 1;
+            thread::sleep(self.config.promote_delay);
+
+            if tries >= self.config.promote_limit {
+                return Err(Error::InvalidTransaction(TransactionError::Unconfirmable));
+            }
+
+            self.promote(&txn).await?;
+        }
+
+        Ok(txn)
+    }
+
+    pub async fn publish_json<T>(&self, did: &DID, data: &T) -> Result<Hash>
+    where
+        T: Serialize,
+    {
+        // Ensure the correct network is selected
+        if !self.network.matches_did(did) {
+            return Err(Error::InvalidDocument(DocumentError::NetworkMismatch));
+        }
+
+        // Build the transfer to push the seriaized data to the tangle.
+        let transfer: Transfer = Transfer {
+            address: create_address(&did)?,
+            value: 0,
+            message: Some(to_string(&data).map_err(core::Error::EncodeJSON)?),
+            tag: None,
+        };
+
+        // Dispatch the transfer to the network
+        let mut bundle: Vec<BundledTransaction> = self.client.send(None).transfers(vec![transfer]).send().await?;
+
+        // Extract the response transaction - there should be exactly one.
```

Comment:
Why should there only be one transaction? If the document has a lot content, the size can exceed a single transaction

Reply:
Agreed, even in Chrysalis P2, where transactions can be pretty massive. You can't make any assumptions.

Reply:
😄  fixed - This was just supposed to grab the tail transaction from the list